framework/include/utils/ADReal.h

```diff
@@ -110,3 +110,15 @@ struct ScalarBinaryOpTraits<MetaPhysicL::DualNumber<V, D, asd>, Real, BinaryOp>
   typedef MetaPhysicL::DualNumber<V, D, asd> ReturnType;
 };
 } // namespace Eigen
+
+namespace Moose
+{
+template <typename T>
+struct ADType;
+
+template <typename T, int M, int N, int O, int M2, int N2>
+struct ADType<Eigen::Matrix<T, M, N, O, M2, N2>>
+{
+  typedef typename Eigen::Matrix<typename ADType<T>::type, M, N, O, M2, N2> type;
```

Comment:
I don't really like splitting the `ADType` specializations. Do you have to have this here? Can you not include `MooseTypes.h` instead? Or perhaps we should move the other specializations here?

Reply:
This file has the Eigen include and all the support necessary for AD support in Eigen. Seems like a better design to have these template specializations near the types they specialize for rather than piling it all in MooseTypes (we use a similar approach with dataLoad/dataStore, don't we?)

Reply:
That just means I didn't catch the piling of Eigen stuff in here before. I think the header name should match what is happening in the file. This is well beyond what `ADReal` means. You could split out the eigen stuff into its own header, like `ADEigen.h` or you could rename this file to something more encompassing